cvprac/cvp_api.py

```diff
@@ -3206,4 +3242,470 @@ def workspace_build_status(self, workspace_id, build_id):
+        '''
+        payload = {
+                "key": {
+                    "id": cc_id
+                },
+                "start": {
+                    "value": True,
+                    "notes": notes
+                }
+        }
+        url = '/api/resources/changecontrol/v1/ChangeControlConfig'
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion >= 6.0:
+                self.log.debug('v6 ' + str(url) + ' ' + str(payload))
+                response = self.clnt.post(url, data=payload, timeout=self.request_timeout)
+                return response
+            elif self.clnt.apiversion < 6.0:
+                    self.log.warning('change_control_start: Change Control Resource APIs are supported'
+                                     ' from 2021.2.0 or newer.')
+                    return None
+        else:
+            return self.clnt.post(url, data=payload, timeout=self.request_timeout)
+
+
+    def change_control_stop(self, cc_id, notes=""):
+        ''' Stop a Change Control using Resource APIs.
+            Supported versions: CVP 2021.2.0 or newer and CVaaS.
+
+            Args:
+                cc_id (string): The ID for the new change control.
+                notes (string): An optional note.
+            Returns:
+                response (dict): A dict that contains...
+                Ex: {"value":{"key":{"id":cc_id}, "start":{"value":false, "notes":"note"}},
+                     "time":"2021-12-14T21:02:21.830306071Z"}
+        '''
+        payload = {
+                "key": {
+                    "id": cc_id
+                },
+                "start": {
+                    "value": False,
+                    "notes": notes
+                }
+        }
+        url = '/api/resources/changecontrol/v1/ChangeControlConfig'
+        # For on-prem check the version as it is only supported from 2021.2.0+
+        if not self.clnt.is_cvaas:
+            if self.clnt.apiversion is None:
+                self.get_cvp_info()
+            if self.clnt.apiversion >= 6.0:
+                self.log.debug('v6 ' + str(url) + ' ' + str(payload))
+                response = self.clnt.post(url, data=payload, timeout=self.request_timeout)
+                return response
+            elif self.clnt.apiversion < 6.0:
+                    self.log.warning('change_control_stop: Change Control Resource APIs are supported'
+                                     ' from 2021.2.0 or newer.')
+                    return None
+        else:
+            return self.clnt.post(url, data=payload, timeout=self.request_timeout)
+
+
+    def change_control_schedule(self, cc_id, schedule_time, notes=""):
+        ''' Schedule a Change Control using Resource APIs.
+            Supported versions: CVP 2021.2.0 or newer and CVaaS.
+
```

Comment:
I got a strange issue on CVP - I am not able to approve a CC that was scheduled using this function, thus the CC cannot execute at the scheduled time. 

I tried the following functions: 
```
cc_id = str(uuid.uuid4())
name = "Test CC gigi9 - schedule"
tasks = ['581']
schedule_time = '2021-12-27T13:44:00.0Z'
print(cc_id)

print(clnt_imp.api.change_control_create_for_tasks(cc_id, name, tasks, series=False))
print(clnt_imp.api.change_control_schedule(cc_id=cc_id, schedule_time=schedule_time, notes='salut'))
print(clnt_imp.api.change_control_approve(cc_id=cc_id, notes='salut', approve=True))
```
Result: 
```
9b75b09b-3019-4905-9801-93b8f4648412
{'value': {'key': {'id': '9b75b09b-3019-4905-9801-93b8f4648412'}, 'change': {'name': 'Test CC gigi9 - schedule', 'rootStageId': 'root', 'stages': {'values': {'root': {'name': 'root', 'rows': {'values': [{'values': ['stage0']}]}}, 'stage0': {'name': 'stage0', 'action': {'name': 'task', 'timeout': 3000, 'args': {'values': {'TaskID': '581'}}}}}}, 'notes': 'randomString'}}, 'time': '2021-12-27T13:09:56.285818464Z'}
{'value': {'key': {'id': '9b75b09b-3019-4905-9801-93b8f4648412'}, 'schedule': {'value': '2021-12-27T13:44:00Z', 'notes': 'salut'}}, 'time': '2021-12-27T13:09:56.352372036Z'}
{'value': {'key': {'id': '9b75b09b-3019-4905-9801-93b8f4648412'}, 'approve': {'value': True, 'notes': 'salut'}, 'version': '2021-12-27T13:09:56.285818464Z'}, 'time': '2021-12-27T13:09:56.451946721Z'}
```

From the UI, the CC is not approved. 
However, the API call is returning that the CC is approved: 
```
print(clnt_imp.api.change_control_approval_get_one(cc_id=cc_id))
{'value': {'key': {'id': '9b75b09b-3019-4905-9801-93b8f4648412'}, 'approve': {'value': True, 'notes': 'salut'}, 'version': '2021-12-27T13:09:56.285818464Z'}, 'time': '2021-12-27T13:09:56.451946721Z'}
```

* From the UI, I tried to approve manually the CC, but this doesn't seem to work. 
* When the CC is supposed to execute, the CC fails with the error message: 
```
CC completed with error: Error executing stage stage0 : {"errorCode":"604","errorMessage":"Invalid request"}
```


Manually creating, scheduling, and approving a CC works without issue. 

Tested on 2021.2.2 CVP server.

Reply:
This seems like a potential bug with the APIs. Or could this potentially be some sort of user created issue with timestamps/timezones/formats? Also I'm interested in the appropriate order or operations here.

Reply:
yes this is unfortunately a bug with the API in CVP, I've added a comment in the example that this might only work from 2021.3.0+; I'll up the apiversion for this, I guess it doesnt' make sense to try to schedule a CC if execution will fail